src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java

```diff
@@ -330,6 +417,258 @@ public void jobFinished () {
+                    amiId,
+                    AMI_CONFIG_PATH);
+                LOG.error(statusMessage);
+                status.fail(statusMessage);
+            }
+        }
+        RunInstancesRequest runInstancesRequest = new RunInstancesRequest()
+                .withNetworkInterfaces(interfaceSpecification)
+                .withInstanceType(instanceType)
+                .withMinCount(count)
+                .withMaxCount(count)
+                .withIamInstanceProfile(new IamInstanceProfileSpecification().withArn(otpServer.ec2Info.iamRoleArn))
+                .withImageId(amiId)
+                .withKeyName(otpServer.ec2Info.keyName)
+                // This will have the instance terminate when it is shut down.
+                .withInstanceInitiatedShutdownBehavior("terminate")
+                .withUserData(Base64.encodeBase64String(userData.getBytes()));
+        final List<Instance> instances = ec2.runInstances(runInstancesRequest).getReservation().getInstances();
+
+        List<String> instanceIds = getIds(instances);
+        Map<String, String> instanceIpAddresses = new HashMap<>();
+        // Wait so that create tags request does not fail because instances not found.
+        try {
+            Waiter<DescribeInstanceStatusRequest> waiter = ec2.waiters().instanceStatusOk();
+            long beginWaiting = System.currentTimeMillis();
+            waiter.run(new WaiterParameters<>(new DescribeInstanceStatusRequest().withInstanceIds(instanceIds)));
+            LOG.info("Instance status is OK after {} ms", (System.currentTimeMillis() - beginWaiting));
+        } catch (Exception e) {
+            LOG.error("Waiter for instance status check failed.", e);
+            status.fail("Waiter for instance status check failed.");
+            // FIXME: Terminate instance???
+            return Collections.EMPTY_LIST;
+        }
+        for (Instance instance : instances) {
+            // The public IP addresses will likely be null at this point because they take a few seconds to initialize.
+            instanceIpAddresses.put(instance.getInstanceId(), instance.getPublicIpAddress());
+            String serverName = String.format("%s %s (%s) %d", deployment.r5 ? "r5" : "otp", deployment.name, dateString, serverCounter++);
+            LOG.info("Creating tags for new EC2 instance {}", serverName);
+            ec2.createTags(new CreateTagsRequest()
+                    .withTags(new Tag("Name", serverName))
+                    .withTags(new Tag("projectId", deployment.projectId))
+                    .withTags(new Tag("deploymentId", deployment.id))
+                    .withTags(new Tag("jobId", this.jobId))
+                    .withTags(new Tag("serverId", otpServer.id))
+                    .withResources(instance.getInstanceId())
+            );
+        }
+        List<Instance> updatedInstances = new ArrayList<>();
+        while (instanceIpAddresses.values().contains(null)) {
```

Comment:
There is no other escape from this loop in the event that some instances don't get assigned public IPs. This is probably unlikely to happen, but it might be a good idea to add some absurd time limit that fails the deploy job just in case.